kentik_api_library/kentik_api/synthetics/api_transport_grpc.py

```diff
@@ -0,0 +1,880 @@
+    pbTestStatus.TEST_STATUS_DELETED: TestStatus.deleted,
+}
+
+PB_TO_IP_FAMILY = {
+    pbIPFamily.IP_FAMILY_UNSPECIFIED: IPFamily.unspecified,
+    pbIPFamily.IP_FAMILY_V4: IPFamily.v4,
+    pbIPFamily.IP_FAMILY_V6: IPFamily.v6,
+    pbIPFamily.IP_FAMILY_DUAL: IPFamily.dual,
+}
+
+PB_TO_AGENT_STATUS = {
+    pbAgentStatus.AGENT_STATUS_UNSPECIFIED: AgentStatus.UNSPECIFIED,
+    pbAgentStatus.AGENT_STATUS_OK: AgentStatus.OK,
+    pbAgentStatus.AGENT_STATUS_WAIT: AgentStatus.WAIT,
+    pbAgentStatus.AGENT_STATUS_DELETED: AgentStatus.DELETED,
+}
+
+PB_TO_AGENT_IMPL_TYPE = {
+    pbAgentImpl.IMPLEMENT_TYPE_UNSPECIFIED: AgentImplementType.UNSPECIFIED,
+    pbAgentImpl.IMPLEMENT_TYPE_RUST: AgentImplementType.RUST,
+    pbAgentImpl.IMPLEMENT_TYPE_NODE: AgentImplementType.NODE,
+}
+
+PB_TO_TASK_STATE = {
+    pbTaskState.TASK_STATE_UNSPECIFIED: TaskState.UNSPECIFIED,
+    pbTaskState.TASK_STATE_CREATED: TaskState.CREATED,
+    pbTaskState.TASK_STATE_UPDATED: TaskState.UPDATED,
+    pbTaskState.TASK_STATE_DELETED: TaskState.DELETED,
+}
+
+
+class SynthGRPCTransport(KentikAPITransport):
+    def __init__(
+        self, credentials: Tuple[str, str], url: str = "synthetics.api.kentik.com:443", proxy: Optional[str] = None
+    ) -> None:
+        (email, token) = credentials
+        self._url = url
+        self._admin = SyntheticsAdminService()
+        self._data = SyntheticsDataService()
+        self._credentials = [
+            ("x-ch-auth-email", email),
+            ("x-ch-auth-api-token", token),
+        ]
+
+    def req(self, op: str, **kwargs) -> Any:
+        OPS: Dict[str, Callable] = {
+            "TestsList": self.tests_list,
+            "TestCreate": self.test_create,
+            "TestGet": self.test_get,
+            "TestPatch": self.test_patch,
+            "TestDelete": self.test_delete,
+            "TestStatusUpdate": self.test_status_update,
+            "AgentsList": self.agent_list,
+            "AgentGet": self.agent_get,
+            "AgentPatch": self.agent_patch,
+            "AgentDelete": self.agent_delete,
+            "GetHealthForTests": self.get_health_for_tests,
+            "GetTraceForTest": self.get_trace_for_test,
+        }
+
+        try:
+            svc = OPS[op]
+        except KeyError:
+            raise RuntimeError(f"Invalid operation '{op}'")
+
+        try:
+            return svc(**kwargs)
+        except RpcError as error:
+            raise new_api_error(error) from error
+
+    def tests_list(self, **kwargs) -> List[SynTest]:
+        tests: List[SynTest] = []
+        pb_tests = self._admin.ListTests(ListTestsRequest(), metadata=self._credentials, target=self._url).tests
+        for pbt in pb_tests:
+            if pbt.type in [TestType.none.value, TestType.bgp_monitor.value]:
+                pb_test = SynTest("[empty]")
+                populate_test_from_pb(pbt, pb_test)
+                tests.append(pb_test)
+            if pbt.type == TestType.mesh.value:
```

Comment:
This whole code should look roughly like this:

```python
def tests_list(self, **kwargs) -> List[SynTest]:
        tests: List[SynTest] = []
        for t in self._admin.ListTests(ListTestsRequest(), metadata=self._credentials, target=self._url).tests:
            test = make_test_from_pb(t)
            if test is None:
                <handle error>
            else:
                tests.append(test)
       return tests
```

The `make_test_from_pb` function may be defined here or in `synth_tests.py` and should look roughly like this:

```python
def make_test_from_pb(pbt: <appropriate_type>) -> Optional[SynTest]:
    cls = {
         <dictionary mapping pb test types to SynTest sub-classes>
    }.get(pbt.type)
    if not cls:
        <log error>
        return None
    return cls.from_pb(pbt)
```

.. just in case you ask: `make_test_from_pb` could be a class method of `SynTest`, but it would complicate imports, because it may (and likely will) lead to circular dependencies.

Reply:
Will consider it while migrating to API 2022

Reply:
added: 
def make_synth_test(testt: pb.Test) -> SynTest:
...